pkg/divert/istio/divert.go

```diff
@@ -0,0 +1,212 @@
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	UPDATE_CONFLICT_RETRIES = 10
+)
+
+// Driver weaver struct for the divert driver
+type Driver struct {
+	Client      kubernetes.Interface
+	IstioClient istioclientset.Interface
+	Manifest    *model.Manifest
+	cache       *divertIstioCache
+}
+
+type divertIstioCache struct {
+	divertVirtualServices    map[string]*istioV1beta1.VirtualService
+	developerVirtualServices map[string]*istioV1beta1.VirtualService
+}
+
+func (d *Driver) Deploy(ctx context.Context) error {
+	for name := range d.cache.divertVirtualServices {
+		select {
+		case <-ctx.Done():
+			oktetoLog.Infof("deployDivert context cancelled")
+			return ctx.Err()
+		default:
+			oktetoLog.Spinner(fmt.Sprintf("Diverting virtual service %s/%s ...", d.Manifest.Deploy.Divert.Namespace, name))
+			if err := d.retryTranslateDivertVirtualService(ctx, name); err != nil {
+				return err
+			}
+
+			if !isInDivertVirtualServices(name, d.Manifest.Deploy.Divert.VirtualServices) {
+				continue
+			}
+			if err := d.retryTranslateDeveloperVirtualService(ctx, name); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *Driver) Destroy(ctx context.Context) error {
+	for name := range d.cache.divertVirtualServices {
+		select {
+		case <-ctx.Done():
+			oktetoLog.Infof("destroyDivert context cancelled")
+			return ctx.Err()
+		default:
+			oktetoLog.Spinner(fmt.Sprintf("Destroying divert un virtual service %s/%s ...", d.Manifest.Deploy.Divert.Namespace, name))
+			if err := d.retryRestoreDivertVirtualService(ctx, name); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (d *Driver) GetDivertNamespace() string {
+	if d.Manifest.Deploy.Divert.Namespace == d.Manifest.Namespace {
+		return ""
+	}
+	return d.Manifest.Deploy.Divert.Namespace
+}
+
+func (d *Driver) InitCache(ctx context.Context) error {
+	d.cache = &divertIstioCache{
+		divertVirtualServices:    map[string]*istioV1beta1.VirtualService{},
+		developerVirtualServices: map[string]*istioV1beta1.VirtualService{},
+	}
+	vsList, err := virtualservices.List(ctx, d.Manifest.Deploy.Divert.Namespace, d.IstioClient)
+	if err != nil {
+		return err
+	}
+	for i := range vsList {
+		d.cache.divertVirtualServices[vsList[i].Name] = vsList[i]
+	}
+
+	vsList, err = virtualservices.List(ctx, d.Manifest.Namespace, d.IstioClient)
+	if err != nil {
+		return err
+	}
+	for i := range vsList {
+		d.cache.developerVirtualServices[vsList[i].Name] = vsList[i]
+	}
+	return nil
+}
+
+func (d *Driver) retryTranslateDeveloperVirtualService(ctx context.Context, name string) error {
+	retries := 0
+	var err error
+	for retries < UPDATE_CONFLICT_RETRIES {
+		if _, ok := d.cache.developerVirtualServices[name]; !ok {
+			vs := createIntoDeveloperVirtualService(d.Manifest, d.cache.divertVirtualServices[name])
+			err = virtualservices.Create(ctx, vs, d.IstioClient)
+			if err == nil {
+				d.cache.developerVirtualServices[name] = vs
+				return nil
```

Comment:
As it is now, if the VS already exists we are not adding it to the cache. From what I understand we should do it and we could merge it with the next `return nil`

Reply:
I removed the whole cache logic. It was complicating the logic and not bringing better performance